File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.js

function solution(today, terms, privacies) {
    var answer = [];
    const dayOfToday = dateToDay(today) // 오늘 날짜
    const daysOfTerms = new Map // 약관 날짜
    for(let i = 0; i<terms.length; i++){
        const termsArr = terms[i].split(" ")
        daysOfTerms.set(termsArr[0], termsArr[1]*28)
    }
    
    for(let i = 0; i < privacies.length; i++){
        const privaciesArr = privacies[i].split(" ")
        const dayOfPrivacies = dateToDay(privaciesArr[0]) // 날짜를 일 수로 변환
        const dayOfTerms = daysOfTerms.get(privaciesArr[1]) // 약관종류 일 수로 변환
        
        if(dayOfPrivacies + dayOfTerms <= dayOfToday) answer.push(i+1)
    }
    
    return answer;
}

function dateToDay(date){
    const dateArr = date.split(".")
    const year = Number(dateArr[0])
    const month = Number(dateArr[1])
    const day = Number(dateArr[2])
    
    return (year-2000)*336 + (month*28) + day
}

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.py
def solution(today, terms, privacies):
    answer = []
    t_y,t_m,t_d = [date for date in today.split('.')]
    today = int(f'{t_y}{t_m}{t_d}')
    
    dic = {}
    for term in terms :
        term_a,term_d = term.split()
        dic[term_a]=int(term_d)
        
    for i,privacy in enumerate(privacies) :
        fulldate, term_a = privacy.split()
        y,m,d = [date for date in fulldate.split('.')]
        m = int(m) + dic[term_a]
        dy,dm = divmod(m,12)
        if not dm :
            dy -= 1
            dm += 12
        y = int(y) + dy
        if dm < 10 :
            m = '0' + str(dm)
        else :
            m = str(dm)

        d_day = int(f'{y}{m}{d}')

        if d_day <= today :
            answer.append(i+1)
    return answer